### PR TITLE
Add native systemd service

### PR DESCRIPTION
**Pull request by MichaIng**
## Thank you for your contribution to the Pi-hole Community! 

Please read the comments below to help us consider your Pull Request.

We are all volunteers and completing the process outlined will help us review your commits quicker.

**Please make sure you**
 1. Base your code and PRs against the repositories developmental branch.
 2. [Sign Off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits as we enforce the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions

---
- **What does this PR aim to accomplish?:**

Adding a native systemd service to avoid the systemd-sysv wrapper units and enhance service status tracking.

- **How does this PR accomplish the above?:**

Inspired by: #2112

A pre-start and a post-stop script are added to reduce doubled setup and cleanup code.

Since systemd services do not natively support dynamic users, test once whether capabilities are supported during install/update, and remove User=pihole otherwise.

- **What documentation changes (if any) are needed to support this PR?:**

I couldn't find a required change. However, if used anywhere, while both works, the `systemctl` command should be preferred over the `service` command for systemd service handling.

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
- [x] I have read the above and my PR is ready for review. _Check this box to confirm_

### CODE REVIEW

File: advanced/Templates/pihole-FTL-poststop.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env sh
+
+# Source utils.sh for getFTLPIDFile(), getFTLAPIPortFile()
+PI_HOLE_SCRIPT_DIR='/opt/pihole'
+utilsfile="${PI_HOLE_SCRIPT_DIR}/utils.sh"
+. "${utilsfile}"
```
Comment (rdwebdesign) [inline advanced/Templates/pihole-FTL-poststop.sh:6]: I think you will need to add `# shellcheck disable=SC2154`
or use the same method used here:
https://github.com/pi-hole/pi-hole/blob/fe8b0e98ab64ae47a6c31eea67cede3cf502d831/pihole#L12-L24

Comment (MichaIng) [reply advanced/Templates/pihole-FTL-poststop.sh:6]: Ah, using `readonly` satisfies shellcheck? How does it find the actual script location, since on a Pi-hole host it has a different path (not only absolute) than on the (checked out) repository.

But `readonly` is a good idea anyway.

Comment (MichaIng) [reply advanced/Templates/pihole-FTL-poststop.sh:7]: Okay, not sure how Stickler CI calls shellcheck, sadly none of the directives to load the actual script works:
```sh
# shellcheck source=../Scripts/utils.sh
# shellcheck source=SCRIPTDIR/../Scripts/utils.sh
# shellcheck source-path=SCRIPTDIR/../Scripts
```
😢. So it's:
```sh
# shellcheck disable=SC1090
```

File: advanced/Templates/pihole-FTL-prestart.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+readonly utilsfile="${PI_HOLE_SCRIPT_DIR}/utils.sh"
+# shellcheck disable=SC1090
+. "${utilsfile}"
+
+# Get file paths
+FTL_PID_FILE="$(getFTLPIDFile)"
+FTL_PORT_FILE="$(getFTLAPIPortFile)"
+
+# Touch files to ensure they exist (create if non-existing, preserve if existing)
+mkdir -pm 0755 /run/pihole /var/log/pihole
```
Comment (MichaIng) [inline advanced/Templates/pihole-FTL-prestart.sh:14]: How to mute/satisfy this Stickler CI error? It is explicitly wanted here to apply modes only to the deepest directory, of course.

Comment (rdwebdesign) [reply advanced/Templates/pihole-FTL-prestart.sh:14]: You can punch it.
stickerCI

Sorry, I don't know.
Stickler is very annoying sometimes.

Comment (MichaIng) [reply advanced/Templates/pihole-FTL-prestart.sh:15]: Found it, shellcheck as well: https://www.shellcheck.net/wiki/SC2174

**Comment by DL6ER** [comment]
Thanks! I have tested this on a fresh Ubuntu server installation where it seems to work as expected:
Screenshot from 2022-09-19 04-36-48
I just noticed that it does pick only a subset of the log output but haven't had the time right now (it's quarter to 5AM) to dig out why:
Screenshot from 2022-09-19 04-38-28

The status is correctly shown even when FTL is killed manually:
Screenshot from 2022-09-19 04-38-08

File: advanced/Templates/pihole-FTL.systemd [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+Before=nss-lookup.target
+
+[Service]
+User=pihole
+PermissionsStartOnly=true
+
+ExecStartPre=/opt/pihole/pihole-FTL-prestart.sh
+ExecStart=/usr/bin/pihole-FTL -f
+Restart=on-abnormal
+RestartSec=30s
```
Comment (DL6ER) [inline advanced/Templates/pihole-FTL.systemd:21]:
```suggestion
Restart=on-failure
RestartSec=5s
```

We should definitely use the `on-failure` policy here to at least *try* to recover from crashes, this isn't the case with `on-abnormal`:
Screenshot from 2022-09-19 04-41-27

> Setting this to `on-failure` is the recommended choice for long-running services, in order to increase reliability by attempting automatic recovery from errors.
> [source](https://www.freedesktop.org/software/systemd/man/systemd.service.html)

This will also make https://github.com/pi-hole/FTL/pull/1377 obsolete which is what triggered(/revived?) the whole idea here.


With `Restart=on-failure` auto-recovery from crash works fine:
Screenshot from 2022-09-19 04-52-00

**Review by DL6ER — CHANGES_REQUESTED** [review]

**Comment by rdwebdesign** [comment]
There is a `-n` option to show more lines:
```sh
-n --lines=INTEGER  Number of journal entries to show
```

Example:
```sh
sudo systemctl status pihole-FTL -n50
```
to see 50 lines.

**Comment by dschaper** [comment]
Or `journalctl -u pihole-FTL`.

File: advanced/Templates/pihole-FTL.systemd (hunk 3 above)
Comment (MichaIng) [reply advanced/Templates/pihole-FTL.systemd:21]: Makes sense: https://www.freedesktop.org/software/systemd/man/systemd.service.html#Restart=

With 5 seconds delay before restarts, the start loop limit interval needs to be increased: https://www.freedesktop.org/software/systemd/man/systemd.unit.html#StartLimitIntervalSec=interval
It's 100ms by default, so Pi-hole could enter an endless restart loop. I'll increase it to 60 seconds so that it aborts after 5 attempts (default) up to 12 seconds each.

File: automated install/basic-install.sh [hunk 4]
```diff
@@ -1381,8 +1381,30 @@ installConfigs() {
...
+    # Install pihole-FTL systemd or init.d service, based on whether systemd is the init system or not
+    # Follow debhelper logic, which checks for /run/systemd/system to derive that systemd is the init system
+    if [[ -d '/run/systemd/system' ]]
+    then
+        install -T -m 0755 "${PI_HOLE_LOCAL_REPO}/advanced/Templates/pihole-FTL.systemd" '/etc/systemd/system/pihole-FTL.service'
+
+        # Set net admin permissions so that FTL can serve DNS, DHCP and IMAP (for DHCPv6). If this does not work, run FTL as root user.
+        if ! setcap CAP_NET_BIND_SERVICE,CAP_NET_RAW,CAP_NET_ADMIN,CAP_SYS_NICE,CAP_IPC_LOCK,CAP_CHOWN+eip '/usr/bin/pihole-FTL'; then
+            sed -i '/^User=/d' '/etc/systemd/system/pihole-FTL.service'
+        fi
```
Comment (MichaIng) [inline automated install/basic-install.sh:1393]: Are you fine with this solution? The init.d service does this dynamically:
```sh
    if setcap CAP_NET_BIND_SERVICE,CAP_NET_RAW,CAP_NET_ADMIN,CAP_SYS_NICE,CAP_IPC_LOCK,CAP_CHOWN+eip "/usr/bin/pihole-FTL"; then
      su -s /bin/sh -c "/usr/bin/pihole-FTL" pihole || exit $?
    else
      echo "Warning: Starting pihole-FTL as root because setting capabilities is not supported on this system"
      /usr/bin/pihole-FTL || exit $?
    fi
```
But this would require another wrapper script which I wanted to avoid.

**Comment by MichaIng** [comment]
Also the migration from init.d to systemd via update seems to work fine, no orphaned process or so. But I could test it only with manual checkout + hacking owner+branch into installer. So I'll do further master > dev update tests once merged.

**Comment by rdwebdesign** [comment]
I have a question:
Will this changes affect the docker version?

**Comment by MichaIng** [comment]
No, since the Docker image does not use systemd. AFAIK it doesn't use SysV either, but with s6 an entirely different init/service system.

File: automated install/basic-install.sh (hunk 4 above)
Comment (DL6ER) [reply automated install/basic-install.sh:1393]: Yeah, I've seen this. I wouldn't expect capabilities support to change at all so defining once (or, rather, every now and then) if `pihole` or `root` should be used seems in no way inferior to doing it on every (re)start.

**Comment by coogor** [comment]
Just FTR: there is a .service file available: https://build.opensuse.org/package/show/home:Herbster0815:HTPC/pi-hole-ftl

**Comment by MichaIng** [comment]
Ah:
```
AmbientCapabilities=CAP_NET_BIND_SERVICE CAP_NET_RAW CAP_NET_ADMIN CAP_SYS_NICE CAP_IPC_LOCK CAP_CHOWN
```
actually is required here as well, for fresh installs with systemd where `setcap` was never applied. I'll add it once I'm back home from weekend trip.

**Comment by yubiuser** [comment]
Needs rebasing after https://github.com/pi-hole/pi-hole/pull/4966 has been merged.

File: advanced/Templates/pihole-FTL.service [hunk 5]
```diff
@@ -28,33 +29,8 @@ start() {
...
-        ln -s /var/log/pihole/pihole.log /var/log/pihole.log
-        chown -h pihole:pihole /var/log/pihole.log
-
-    fi
-    if [ ! -f /var/log/pihole-FTL.log ]; then
-        ln -s /var/log/pihole/FTL.log /var/log/pihole-FTL.log
-        chown -h pihole:pihole /var/log/pihole-FTL.log
-    fi
+    # Run pre-start script, which pre-creates all expected files with correct permissions
+    "${PI_HOLE_SCRIPT_DIR}/pihole-FTL-prestart.sh"
```
Comment (DL6ER) [inline advanced/Templates/pihole-FTL.service:33]:
```suggestion
    sh "${PI_HOLE_SCRIPT_DIR}/pihole-FTL-prestart.sh"
```

File: advanced/Templates/pihole-FTL.service [hunk 6]
```diff
@@ -89,8 +65,8 @@ stop() {
   else
     echo "Not running"
   fi
-  # Cleanup
-  rm -f /run/pihole/FTL.sock /dev/shm/FTL-* "${FTL_PID_FILE}"
+  # Run post-stop script, which does cleanup among runtime files
+  "${PI_HOLE_SCRIPT_DIR}/pihole-FTL-poststop.sh"
```
Comment (DL6ER) [inline advanced/Templates/pihole-FTL.service:69]:
```suggestion
  sh "${PI_HOLE_SCRIPT_DIR}/pihole-FTL-poststop.sh"
```

File: automated install/basic-install.sh [hunk 7]
```diff
@@ -1377,8 +1377,30 @@ installConfigs() {
         fi
     fi
 
-    # Install pihole-FTL.service
-    install -T -m 0755 "${PI_HOLE_LOCAL_REPO}/advanced/Templates/pihole-FTL.service" "/etc/init.d/pihole-FTL"
+    # Install pihole-FTL systemd or init.d service, based on whether systemd is the init system or not
+    # Follow debhelper logic, which checks for /run/systemd/system to derive whether systemd is the init system
+    if [[ -d '/run/systemd/system' ]]
+    then
```
Comment (DL6ER) [inline automated install/basic-install.sh:1383]:
```suggestion
    if [[ -d '/run/systemd/system' ]]; then
```
just to match the style used everywhere else in this file

**Review by DL6ER — COMMENTED** [review]
Thanks again for your PR. I checked again and everything seems to work just fine. I added a few minor comments (like adding `sh` to have to less rely on the `x` flag) and then we should merge it into `development` branch for broader testing.

File: advanced/Templates/pihole-FTL.service (hunk 5 above)
Comment (MichaIng) [reply advanced/Templates/pihole-FTL.service:33]: Isn't this redundant as of the shebang of the scripts (same with poststop)? I'm not aware of any difference between calling a script with a shell explicitly and calling the same shell via shebang instead.

Comment (MichaIng) [reply advanced/Templates/pihole-FTL.service:33]: Ah, this makes sense. A counter argument is: If the permissions of a file really changed, the admin should be probably made ware of, instead of ignoring it. Possibly the whole content of the file got accidentally or with bad intention changed, and we call a dangerous script then.

But for this, root permissions would have been required already, so less me max flexible/compatible for now.

**Review by DL6ER — APPROVED** [review]

**Comment by PromoFaux** [comment]
> but with s6 an entirely different init/service system.

It's a dark art, but it's more or less a copy of the existing service

https://github.com/pi-hole/docker-pi-hole/blob/master/src/s6/debian-root/etc/s6-overlay/s6-rc.d/pihole-FTL/run

**Comment by MichaIng** [comment]
> it's more or less a copy of the existing service

It could use `pihole-FTL-prestart.sh`/`pihole-FTL-poststop.sh` scripts then as well, when this got merged, to reduce maintenance a bit.

**Review by yubiuser — COMMENTED** [review]
Currently, all tests pass. However, they will fail once we add systemd to one of our test containers (e.g. https://github.com/pi-hole/pi-hole/pull/5027)  because of

https://github.com/pi-hole/pi-hole/blob/21026d941428bfcb4f405479db470a90bae6aec1/test/test_any_automated_install.py#L279-L284

I implemented this code to circumvent the test if it is not applicable:

https://github.com/pi-hole/pi-hole/pull/5027/commits/83f983f4e8885f5e22ad3a96ba16bdc56ba5ea01

File: automated install/basic-install.sh [hunk 8]
```diff
@@ -1377,8 +1377,29 @@ installConfigs() {
...
-    install -T -m 0755 "${PI_HOLE_LOCAL_REPO}/advanced/Templates/pihole-FTL.service" "/etc/init.d/pihole-FTL"
+    # Install pihole-FTL systemd or init.d service, based on whether systemd is the init system or not
+    # Follow debhelper logic, which checks for /run/systemd/system to derive whether systemd is the init system
+    if [[ -d '/run/systemd/system' ]]; then
+        install -T -m 0644 "${PI_HOLE_LOCAL_REPO}/advanced/Templates/pihole-FTL.systemd" '/etc/systemd/system/pihole-FTL.service'
+
+        # Set net admin permissions so that FTL can serve DNS, DHCP and IMAP (for DHCPv6). If this does not work, run FTL as root user.
+        if ! setcap CAP_NET_BIND_SERVICE,CAP_NET_RAW,CAP_NET_ADMIN,CAP_SYS_NICE,CAP_IPC_LOCK,CAP_CHOWN+eip '/usr/bin/pihole-FTL'; then
+            sed -i '/^User=/d' '/etc/systemd/system/pihole-FTL.service'
+        fi
```
Comment (MichaIng) [inline automated install/basic-install.sh:1388]: I just realized: With `AmbientCapabilities=` in the systemd unit, setting capabilities on the binary is redundant, isn't it? The init.d script still does this internally, like it did before.

Comment (yubiuser) [reply automated install/basic-install.sh:1388]: What happens if we remove this part and `systemd` fails to set `AmbientCapabilities=`? Do we need a check  for this so FTL can run as `root`?

Comment (MichaIng) [reply automated install/basic-install.sh:1388]: Well, this is not a command one can test but a native systemd feature. If systemd fails to apply directives, the whole service fails (usually). Since a lot of systemd services shipped with systemd itself and other packages use this feature, the system would be broken if this didn't work, so IMO it's not reasonable to check for such basic systemd features. But I also don't know about a case where `setcap` fails (aside of missing permissions, R/O, immutable flag and such things, of course, which is then a mis-configuration on a different level).

@DL6ER since you did #2378
Do you remember the cases where this happened, and are such systems/kernels even able to run systemd as init system, or is it narrowed down custom build for embedded systems which run other init systems or only single applications at boot anyway?

Comment (DL6ER) [reply automated install/basic-install.sh:1388]: The issue really isn't with old kernels (the last 2.4 was released 2011 - such devices really shouldn't be connected to the Internet these days) but rather the other aspect: file systems. Capabilities are dependent on the file system supporting extended attributes (like Ext2, Ext3, Ext4, Btrfs, JFS, XFS, and Reiserfs) but many popular ones used by proprietary network file systems (FAT, NTFS) do not.

I think we discovered it back then because `docker` suffers (or suffer*ed* ?) from the issue that, when a volume is stored on these filesystems, capabilities weren't available either. I also remember some users reporting errors running Pi-hole on cheepy VPS where setting capabilities was prohibited for the user machines - either as a "security" measure or, more likely, because the administrators don't understand what capabilities are good for and that they can never grant you more rights than your `root` user has anyway.

Having said all that, we should also remember that the PR you mentioned is more than 4 years old. Some things might have changed but surely not all. I'd consider `AmbientCapabilities` to be the superior choice in terms of compatibility, however, the currently used hack of starting as `root` and having FTL recognize this and drop itself to user `pihole` once it did what it needs `root`-powers for (binding to low-number ports for DNS; DHCP, etc.) works amazingly well, too.

TL;DR: `systemd` is using [`libcap`](https://linux.die.net/man/3/libcap) to grant capabilities to the just started *process*. You need to be `root` to do that (`sudo systemctl ...`). This is a different approach than ours - we are adding these capabilities to the binary and rely on the kernel to see this when executing the binary.

Comment (MichaIng) [reply automated install/basic-install.sh:1388]: Okay, FAT and NTFS cannot be used as rootfs anyway, but in (probably older) Docker or container images, e.g. with aufs instead of overlayfs, on filesystem basis this probably still makes sense. But in a container, you won't run systemd, and the systemd `AmbientCapabilities` does not use capabilities on filesystem level, so I think we can remove the check and fallback to root user in systemd unit.

But it still makes sense to (try) `setcup` the binary in addition to `AmbientCapabilities`, so that manual FTP starts, for testing, debugging or whatever, remains possible without doing it as root or granting capabilities first.
